subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,698 @@
+	if ch.currentValue == nil {
+		return "", newEmptyBluetoothCharacteristicError(c)
+	}
+	return *ch.currentValue, nil
+}
+
+// initializeWriteOnlyBluetoothCharacteristic returns a bluetooth characteristic config.
+func initializeWriteOnlyBluetoothCharacteristic(bsl *bluetoothServiceLinux, cName string, encoding uint16,
+) bluetooth.CharacteristicConfig {
+	cUUID := bluetooth.NewUUID(uuid.New()).Replace16BitComponent(encoding)
+	bsl.logger.Debugf("%s can be written to the following bluetooth characteristic: %s", cName, cUUID.String())
+
+	// characteristic represents the in-memory storage for the characteristic identifiable by cName.
+	characteristic := &bluetoothCharacteristicLinux[*string]{
+		UUID:         cUUID,
+		currentValue: nil,
+	}
+	bsl.characteristicsByName[cName] = characteristic
+
+	return bluetooth.CharacteristicConfig{
+		UUID:  cUUID,
+		Flags: bluetooth.CharacteristicWritePermission,
+
+		// WriteEvent is defined by the bluetooth package and is the callback function which handles
+		// write events to bluetooth characteristics. Once written, it is our job to "save" these
+		// values to in-memory storage.
+		WriteEvent: func(client bluetooth.Connection, offset int, value []byte) {
+			v := string(value)
+			bsl.logger.Infof("Received %s: %s from client with connection ID: %d", cName, v, client)
+
+			// Mutex locks each characteristic that is being written so simultaneous calls to
+			// bsl.readCharacteristic(cName) safely access the characteristic value.
+			characteristic.mu.Lock()
+			defer characteristic.mu.Unlock()
+			characteristic.currentValue = &v
+		},
+	}
+}
+
+// initializeReadOnlyBluetoothCharacteristic returns a bluetooth characteristic config.
+func initializeReadOnlyBluetoothCharacteristic(bsl *bluetoothServiceLinux, cName string, encoding uint16) bluetooth.CharacteristicConfig {
+	cUUID := bluetooth.NewUUID(uuid.New()).Replace16BitComponent(encoding)
+	bsl.logger.Debugf("%s can be read from the following bluetooth characteristic: %s", cName, cUUID.String())
+	c := &bluetooth.Characteristic{}
+
+	// refreshAvailableWiFiNetworks represents the callback function used
+	// to dynamically update an otherwise read-only bluetooth characteristic.
+	bsl.refreshAvailableWiFiNetworks = func(bs []byte) error {
+		_, err := c.Write(bs)
+		return err
+	}
+	return bluetooth.CharacteristicConfig{
+		Handle: c,
+		UUID:   cUUID,
+		Flags:  bluetooth.CharacteristicReadPermission,
+	}
+}
+
+// initializeBluetoothService performs low-level system configuration to enable bluetooth advertisement.
+func initializeBluetoothService(deviceName string, characteristics []bluetooth.CharacteristicConfig,
+) (bluetooth.UUID, *bluetooth.Advertisement, error) {
+	adapter := bluetooth.DefaultAdapter
+	if err := adapter.Enable(); err != nil {
+		return [4]uint32{}, nil, fmt.Errorf("failed to enable bluetooth adapter: %w", err)
+	}
+	serviceUUID := bluetooth.NewUUID(uuid.New()).Replace16BitComponent(0x1111)
```

Comment:
Question: I see these UUIDs getting used all over, but uuid.New() is generating NEW uuids every run of the software. Is that expected/normal? Or are you suppose to define the full thing once/permanantly for a "known service" or anything like that? Just feels weird to generate new UUIDs for consistent elements.